File: 一轉示範/多執行緒 銀行  C#示範/Bank.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace 多執行緒_銀行__示範
{
    class Bank
    {
        public int money { get; set; } = 10000;
        public int outlay { get; set; } = 0;  //放款
        public int income { get; set; } = 10000;  //收入
        private bool _stop = false;
        public bool stop
        {
            get
            {
                if (money <= 0)  //沒錢就停
                    _stop = true;
                return _stop;
            }

            set
            {
                _stop = value;
            }
        }


        public void deposit(int money)
        {
            lock (this)
            {
                this.money += money;
                stop = false;
                income += money;
            }
        }

        public int draw(int money)
        {
            lock (this)
            {
                int result;
                if (this.money < money)
                    result = this.money;
                else
                    result = money;
                this.money -= result;

                outlay += result;

                return result;
            }
            
        }

    }
}

File: 一轉示範/多執行緒 銀行示範/src/Bank.java

public class Bank {
	private static final int DELAY = 100;
	private int money = 10000;
	private int outlay = 0;
	private int income = 10000;
	private boolean isBroke = false;
	
	public synchronized void deposit(User user, int depositeMoney) throws InterruptedException{
		if (!isBroke())
		{
			this.money += depositeMoney;
			income += depositeMoney;
			System.out.printf("User %s is depositting , Amount : %d ...%n" , Thread.currentThread().getName() , depositeMoney);
			Thread.sleep(DELAY);
		}
		
	}
	
	public synchronized int draw(User user, int requestMoney) throws InterruptedException {
		if (!isBroke())
		{
			int result  = this.money < requestMoney ? this.money : requestMoney;
			this.money -= result;
			outlay += result;
			System.out.printf("User %s is drawing , Amount : %d ... , Get : %d %n" , Thread.currentThread().getName() , requestMoney , result);
			Thread.sleep(DELAY);
			return result; 
		}
		
		return 0;
	}
	
	public  boolean isBroke(){
		return this.money <= 0; 
	}

	public int getOutlay() {
		return outlay;
	}

	public int getIncome() {
		return income;
	}

	public void setIncome(int income) {
		this.income = income;
	}
	
	public int getMoney() {
		return money;
	}
	
}
